### PR TITLE
Finesse Brave Ads opportunity P2A; Remove redundant code; Cleanup unit tests

### PR DESCRIPTION
**Pull request by tmancey**
Resolves https://github.com/brave/brave-browser/issues/32880

## Submitter Checklist:

- [x] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

General testing of text classification, purchase intent and P2A.

### CODE REVIEW

**Review by goodov — APPROVED** [review]
`chromium_src` lgtm

**Review by kylehickinson — APPROVED** [review]
iOS++

File: components/brave_ads/browser/analytics/p2a/p2a.cc [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+std::string GetStoragePath(const std::string& name) {
+  return base::StrCat({prefs::kP2AStoragePrefNamePrefix, name});
+}
+
+void EmitP2AHistogramName(const std::string& name, const uint64_t sum) {
+  CHECK(base::Contains(kAllowedNames, name));
+
+  const size_t* const iter =
+      std::lower_bound(kIntervalBuckets, std::cend(kIntervalBuckets), sum);
+  const size_t bucket = iter - kIntervalBuckets;
```
Comment (aseren) [inline components/brave_ads/browser/analytics/p2a/p2a.cc:33]: std::distance(std::cbegin(kIntervalBuckets), iter);

Comment (tmancey) [reply components/brave_ads/browser/analytics/p2a/p2a.cc]: Changed, thanks

File: components/brave_ads/browser/analytics/p2a/p2a.cc [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+std::string GetStoragePath(const std::string& name) {
+  return base::StrCat({prefs::kP2AStoragePrefNamePrefix, name});
+}
+
+void EmitP2AHistogramName(const std::string& name, const uint64_t sum) {
+  CHECK(base::Contains(kAllowedNames, name));
+
+  const size_t* const iter =
+      std::lower_bound(kIntervalBuckets, std::cend(kIntervalBuckets), sum);
```
[the remark this answers is not in the document]
Comment (tmancey) [reply components/brave_ads/browser/analytics/p2a/p2a.cc]: Changed, thanks

File: components/brave_ads/core/internal/analytics/p2a/impressions/p2a_impression_util.cc [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+namespace brave_ads {
+
+namespace {
+
+constexpr char kAdimpressionsPerSegmentEvent[] =
+    "Brave.P2A.$1.impressions_per_segment.$2";
+constexpr char kAdImpressionsEvent[] = "Brave.P2A.$1.impressions";
+
+absl::optional<std::string> BuildAdimpressionsPerSegmentEvent(
```
Comment (szilardszaloki) [inline components/brave_ads/core/internal/analytics/p2a/impressions/p2a_impression_util.cc:22]: Nit: `BuildAdImpressionsPerSegmentEvent`

File: components/brave_ads/core/internal/analytics/p2a/opportunities/p2a_opportunity_util.cc [hunk 4]
```diff
@@ -5,53 +5,64 @@
...
-std::string NormalizeSegment(const std::string& segment) {
-  std::string normalized_segment;
-  base::ReplaceChars(StripNonAlphaNumericCharacters(segment), " ", "",
-                     &normalized_segment);
-
-  return normalized_segment;
-}
-
-absl::optional<std::string> BuildAdOpportunitiesPerSegmentEvent(
+absl::optional<std::string> BuildAdopportunitiesPerSegmentEvent(
```
Comment (szilardszaloki) [inline components/brave_ads/core/internal/analytics/p2a/opportunities/p2a_opportunity_util.cc:22]: Nit: `BuildAdOpportunitiesPerSegmentEvent`

File: components/brave_ads/core/internal/analytics/p2a/impressions/p2a_impression_util.cc (hunk 3 above)
Comment (tmancey) [reply components/brave_ads/core/internal/analytics/p2a/impressions/p2a_impression_util.cc]: Fixed

File: components/brave_ads/core/internal/serving/ad_serving_util.h [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+#include "base/check.h"
+#include "base/rand_util.h"
+
+namespace brave_ads {
+
+template <typename T>
+T ChooseCreativeAd(const std::vector<T>& creative_ads) {
+  CHECK(!creative_ads.empty());
+
+  const int rand = base::RandInt(0, static_cast<int>(creative_ads.size()) - 1);
```
Comment (szilardszaloki) [inline components/brave_ads/core/internal/serving/ad_serving_util.h:20]: Probably `base::RandGenerator(creative_ads.size())` is better, as you're passing `rand` to a `std::vector<>::size_type`.

Comment (tmancey) [reply components/brave_ads/core/internal/serving/ad_serving_util.h]: Changed

File: components/brave_ads/browser/analytics/p2a/p2a.cc [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+  weekly_storage.AddDelta(1);
+
+  EmitP2AHistogramName(name, weekly_storage.GetWeeklySum());
+}
+
+void SuspendP2AHistograms() {
+  // Record a special value to prevent sending this week's data to the P2A
+  // server. Matches |kSuspendedMetricValue| in |p3a_service.cc|.
+  for (const char* const name : kAllowedNames) {
+    base::UmaHistogramExactLinear(name, INT_MAX,
```
Comment (aseren) [inline components/brave_ads/browser/analytics/p2a/p2a.cc:75]: std::numeric_limits<int>::max()

File: components/brave_ads/core/internal/segments/segment_util.cc [hunk 7]
```diff
@@ -122,4 +124,9 @@ bool ShouldFilterSegment(const std::string& segment) {
   return iter != filtered_segments.cend();
 }
 
+void NormalizeSegment(std::string& segment) {
+  base::ReplaceChars(StripNonAlphaNumericCharacters(segment), " ", "",
+                     &segment);
+}
```
Comment (szilardszaloki) [inline components/brave_ads/core/internal/segments/segment_util.cc:130]: We talked about that one — just making sure we don't forget to revert this (as the input and output here are not necessarily of the same size).

**Review by szilardszaloki — APPROVED** [review]

File: components/brave_ads/browser/analytics/p2a/p2a.cc (hunk 6 above)
Comment (tmancey) [reply components/brave_ads/browser/analytics/p2a/p2a.cc]: Changed, thanks

**Review by aseren — APPROVED** [review]
LGTM

**Review by DJAndries — APPROVED** [review]
p3a ++

File: components/brave_ads/core/internal/segments/segment_util.cc (hunk 7 above)
Comment (tmancey) [reply components/brave_ads/core/internal/segments/segment_util.cc]: Reverted as discussed

**Pull request merged**